### ISSUE TITLE
Resolve relative links correctly

### ISSUE THREAD

**Issue by iaindillingham** [opensafely-core/actions-registry#54]
If you visit <https://actions.opensafely.org/actions/cohort-report/v2.1.0/>, then you'll see a link to *DEVELOPERS.md*. That's a relative link, but the file is hosted on GitHub. Consequently, the server responds with a 404.

**Comment by sebbacon** [comment]
[Another example](https://actions.opensafely.org/actions/deciles-charts/v0.0.15/)

![image](https://user-images.githubusercontent.com/211271/166445436-ebb2224f-7e9c-407b-8564-68fb8ff18970.png)

**Comment by CarolineMorton** [comment]
Looking at this, we have 3 options:

1. **Change all the relative links to absolute links in Github**. => because they are absolute, they should still route correctly if clicking from both Github and the actions.opensafely.org website. 

    - Advantage: Quick and easy
    - Disadvantage: 1)We would need to ensure that any authors of new actions knew that the link in the readme had to be an absolute link. 2) The link still goes out to Github. This may look less pretty and fancy to a new user or someone interested in the action and has not heard of Github.

2. **Parse and alter the links** => We could create a function in the actions-registry that when the readme is passed through the HTML to be rendered, the link is altered to be an absolute link
    - Advantage: Should be relatively quick to do this; We could write a test for function.
    - Disadvantage: Same as above- the link still goes out to Github.
    
3. **Render the developer.md in the actions.opensafely.org** => We could reuse the code that we are using to parse the readme markdown into html, to parse the markdown from developers.md. This could be hosted at something like: `https://actions.opensafely.org/actions/<action_name>/developer_notes`. 

    - Advantage: Takes longer to implement. We can apply nice css to the readme and make it look better especially for the user who is not familiar with Github. 
    - Disadvantage: Longer to implement, and additional URL (i.e./developer_notes`). 

Note in all of these solutions, we do not deal with versioning if the developer.md changes. What are people's thoughts on this?

**Comment by CarolineMorton** [comment]
Personally, I think option 3 is the way to go.

**Comment by iaindillingham** [comment]
Option 1 is neither quick nor easy, IMHO 🙂 Both GH and registry URLs contain version information (SHA, branch, tag, etc.) and would involve taking the user from one location to the other location (as you state).

I can't remember how we display the README in the registry, but could we not generalize to handle alternative file types? I think this is your Option 3, but for more than just Markdown files.

**Comment by sebbacon** [comment]
I guess there's also an option (4): require that README.md is entirely self-contained, e.g. screenshots are embedded; and that every project has DEVELOPERS.md that is linked to programmatically from the actions.opensafely.org chrome, so it's signposted as techy (e.g. with a github icon and link text "detailed information for developers (on github).

Or similar.

**Comment by alarthast** [comment]
Catching up on the conversation above, it seems like the Advantages/Disadvantages listed mainly apply to the developer docs (or similar markdown files). The same approach and concerns need not apply to images

For images, I believe that we should avoid displaying the alt text in the first place and just show the image. Option 2 is an easy way to do this. The downside does not apply here since the image will be shown and the user need not be sent somewhere else. This also avoids requiring that screenshots be embedded.

https://github.com/opensafely-core/actions-registry/pull/402 implements Option 2 for images only.

**Comment by iaindillingham** [comment]
I'm afraid I don’t understand your thoughts, Alice.[^1] We want to resolve images and links correctly. I believe your PR resolves images, but not links. We should display the alt text as one would expect alt text to be displayed: as the value of [the `alt` attribute][1].

[^1]: https://bennettoxford.slack.com/archives/C08EK42PTL6/p1741356834635829

[1]: https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt#usage_notes

**Comment by alarthast** [comment]
Indeed, the PR above resolves only images, not all links. I think I accidentally expected mind-reading in my comment above - I'll try to rephrase. I'm proposing that image links and text links be treated differently.

- Image links
  - The markdown version of the readme on github shows the image, not the alt text([example](https://github.com/opensafely-actions/deciles-charts/)). It's only when we pull the readme html from github and put it on the actions registry site that these images are shown as (currently broken) alt-text links.
  - I'm not sure why we should require a visitor on the actions registry site to click into a link to see an image. Would it perhaps not be simpler for the user to just see the image on the page instead of the alt text as a link? (Wrt the above comment, I'm not sure why one would expect alt text to be displayed, I'm afraid ...)
  - Therefore, I propose to just alter the links for the images so that they become absolute urls, so that when a visitor goes to the actions registry site, the images will display. i.e. I think for image links, Option 2 is actually a very good course of action to take. 

- Text links
  - Resolving text links have a different set of concerns to image links, in my opinion. Text links are something whose sole purpose is to send the user somewhere else, unlike image links whose primary end goal is for the user to see the image (imo, at least). 
  - Text links therefore come with a set of concerns, the primary one being whether we should send the user to github or send them to another url within actions registry.
  - If we want to avoid just sending users to github since it looks too technical, Option 2 is not a good course of action to take, Option 4 (having some sort of signposting) or Option 3 (creating another view in action registry) are both better than Option 2 in that regard.

I believe there is no necessity to adopt the same link-resolving solution for both image and text links, so I suggest splitting this ticket into two pieces of work, i.e. two PRs (or even two sub-issues). The PR linked above is proposed just for the image links part.

**Comment by iaindillingham** [comment]
Thanks for clarifying, @alarthast.

Whilst this issue could be two issues -- one for link tags (`<a ...>`) and one for image tags (`<img ...>`) -- I think the mechanism is the same in both cases, isn't it? The mechanism should transform relative URLs into absolute URLs that point to GitHub. It shouldn't matter whether relative URLs appear as the value of `href` or `src` attributes.

The implication is that links will point from OS Actions to GitHub, meaning that a user will go from the former site to the latter site. Images will be displayed on OS Actions, but will be served from GitHub.

Notice that when a browser can't render an image because it can't find the value of the `src` attribute, it doesn't replace the image tag for a link tag.

**Comment by alarthast** [comment]
Thank you Iain; indeed the mechanism to transform github relative urls to github absolute urls would be the same for both. 

What I've gathered from the above conversation is that two years ago, we had some reservations about sending users to github when they click on a link, because it might look too technical. If that's no longer a concern, we can use the same mechanism for all github relative links, image or not. 

But perhaps we would like to implement something in the spirit of Options 3 or 4? (E.g. Caroline's original Option 3 proposes creating a specific view/url for developer docs, which seems doable and need not be done for images.)

**Comment by alarthast** [comment]
Actually my mistake, no, the mechanism is slightly different for images and other urls.

Images require the "raw" content url to work as the `src` attribute of a `<img>` tag: 
- This works: https://raw.githubusercontent.com/opensafely-actions/deciles-charts/v0.0.46/img/deciles_chart_has_sbp_event_by_stp_code.png 
- So does this: https://github.com/opensafely-actions/deciles-charts/raw/main/raw/img/deciles_chart_has_sbp_event_by_stp_code.png
- And this: https://github.com/opensafely-actions/deciles-charts/blob/main/raw/img/deciles_chart_has_sbp_event_by_stp_code.png?raw=true
- But not this: https://github.com/opensafely-actions/deciles-charts/blob/v0.0.46/img/deciles_chart_has_sbp_event_by_stp_code.png

Whereas we certainly don't want to be setting raw=true for something like a markdown file (i.e. we want to link to https://github.com/opensafely-actions/deciles-charts/blob/v0.0.46/DEVELOPERS.md, not https://github.com/opensafely-actions/deciles-charts/blob/v0.0.46/DEVELOPERS.md?raw=true)

It's probably just implementing an if/else statement to differentiate the two, but the difference is not zero.

**Comment by iaindillingham** [comment]
At some point, we may have wondered whether GitHub's UI would confuse a user. Let's not wonder about that user any more. Every user of OpenSAFELY will, at some stage, encounter GitHub's UI. Of course, we may wonder about a user of OpenSAFELY Actions who isn't a user of OpenSAFELY; a member of the public, for example. Let's not worry about that user, either. There's so much more to confuse them than OpenSAFELY Actions.

So, and as you say, we need to differentiate between image and link tags, not just because the attributes are different, but because the URLs are slightly different, too.

**Comment by alarthast** [comment]
Thank you Iain, in that case only one PR is needed to close this ticket, which I'll work on now.
